### PR TITLE
fix(warn): better message with no constructors props

### LINKED ISSUE #9224: Warn when the type of a prop isn't a constructor
### What problem does this feature solve?
It suppose better error handling for prop types errors.
Steps to reproduce error:
1. Set prop type equals to string ('String', not String):
```javascript
exampleProp: {
  type: 'String',
}
```

2. Run project and get an error:
```javascript
[Vue warn]: Error in nextTick: "TypeError: Right-hand side of 'instanceof' is not an object"
TypeError: Right-hand side of 'instanceof' is not an object
    at assertType (vue.esm.js:1668)
    at assertProp (vue.esm.js:1626)
    at validateProp (vue.esm.js:1560)
    at loop (vue.esm.js:3334)
    at initProps (vue.esm.js:3367)
    at initState (vue.esm.js:3308)
    at VueComponent.Vue._init (vue.esm.js:4628)
    at new VueComponent (vue.esm.js:4798)
    at createComponentInstanceForVnode (vue.esm.js:4310)
    at init (vue.esm.js:4131)
```

Error log is not evident and it is difficult to understand the source of problem (first time with this error anyway).

### What does the proposed API look like?
Error message in this case may look like:
```javascript
"Incorrect prop type 'exampleProp' at:123 (where 123 is example string position in the code)"
```

### PR DESCRIPTION
**Pull request by Zarad1993**
Closes #9224

**What kind of change does this PR introduce?** (check at least one)

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update
- [ ] Refactor
- [ ] Build-related changes
- [x] Other, please describe: Error Handling

**Does this PR introduce a breaking change?** (check one)

- [ ] Yes
- [x] No

If yes, please describe the impact and migration path for existing applications:

**The PR fulfills these requirements:**

- [x] It's submitted to the `dev` branch for v2.x (or to a previous version branch), _not_ the `master` branch
- [x] When resolving a specific issue, it's referenced in the PR's title (e.g. `fix #xxx[,#xxx]`, where "xxx" is the issue number)
- [x] All tests are passing: https://github.com/vuejs/vue/blob/dev/.github/CONTRIBUTING.md#development-setup
- [x] New/updated tests are included

If adding a **new feature**, the PR's description includes:
- [ ] A convincing reason for adding this feature (to avoid wasting your time, it's best to open a suggestion issue first and wait for approval before working on it)

**Other information:**
### Proof
**Before**

screen shot 2018-12-22 at 1 36 46 pm

**After**

screen shot 2018-12-22 at 1 37 43 pm

### CODE REVIEW

**Comment by Zarad1993** [comment]
@posva - Do you think this needs any changes or do you think it might be promoted to code review?

**Comment by Zarad1993** [comment]
@posva - Just following up on this. A friendly reminder to have a peak at it when possible 😄

File: src/core/util/props.js [hunk 1]
```diff
@@ -122,13 +122,14 @@ function assertProp (
...
     for (let i = 0; i < type.length && !valid; i++) {
-      const assertedType = assertType(value, type[i])
+      const assertedType = assertType(value, type[i], vm)
       expectedTypes.push(assertedType.expectedType || '')
       valid = assertedType.valid
     }
   }
 
-  if (!valid) {
+  const haveExpectedTypes = expectedTypes.filter(t => t).length;
```
Comment (sid-maddy) [inline src/core/util/props.js:131]:
```suggestion
  const haveExpectedTypes = expectedTypes.some(t => t)
```
Also, remove the semicolons to adhere to the codestyle?

File: src/core/util/props.js [hunk 2]
```diff
@@ -166,7 +167,12 @@ function assertType (value: any, type: Function): {
   } else if (expectedType === 'Array') {
     valid = Array.isArray(value)
   } else {
-    valid = value instanceof type
+    try {
+      valid = value instanceof type
+    } catch (e) {
+      warn('Invalid prop type: "' + String(type) + '" is not a constructor', vm);
+      valid = false;
```
Comment (sid-maddy) [inline src/core/util/props.js:174]:
```suggestion
      warn(`Invalid prop type: "${type}" is not a constructor`, vm)
      valid = false
```
Is the `String` call necessary?

Comment (posva) [reply src/core/util/props.js:174]: Using String is more robust while the interpolation fails for some types like Symbol (which doesn't apply here)

**Review by posva — APPROVED** [review]

**Pull request merged**